File: cpp/src/parser.cpp
// AST construction. Port of rust/crates/agentgate-core/src/parser.rs.

#include "agentgate/agentgate.hpp"

namespace agentgate {
namespace {

struct Parser {
  const std::vector<Tok>& t;
  std::size_t i = 0;
  std::string error;

  explicit Parser(const std::vector<Tok>& toks) : t(toks) {}

  bool fail(const char* msg) {
    if (error.empty()) error = msg;
    return false;
  }

  bool eof() const { return i >= t.size(); }
  const Tok* peek() const { return eof() ? nullptr : &t[i]; }
  const Tok* at(std::size_t k) const { return k < t.size() ? &t[k] : nullptr; }

  bool at_separator() const {
    const Tok* p = peek();
    if (p == nullptr) return false;
    return p->kind == Tok::Kind::Semi || p->kind == Tok::Kind::Amp ||
           p->kind == Tok::Kind::Newline || p->kind == Tok::Kind::AndIf ||
           p->kind == Tok::Kind::OrIf;
  }

  void skip_newlines() {
    while (!eof() && t[i].kind == Tok::Kind::Newline) ++i;
  }

  bool at_stop(bool has_stop, Tok::Kind stop) const {
    if (!has_stop) return false;
    const Tok* p = peek();
    return p != nullptr && p->kind == stop;
  }

  // NAME=value at the head of a word.
  static bool try_split_assignment(const Word& w, std::string& name, Word& value) {
    if (w.segs.empty()) return false;
    if (w.segs[0].kind != Segment::Kind::Lit) return false;
    const std::string& first = w.segs[0].text;
    const std::size_t eq = first.find('=');
    if (eq == std::string::npos || eq == 0) return false;
    const std::string candidate = first.substr(0, eq);
    const char c0 = candidate[0];
    const bool alpha = (c0 >= 'a' && c0 <= 'z') || (c0 >= 'A' && c0 <= 'Z') || c0 == '_';
    if (!alpha) return false;
    for (std::size_t k = 1; k < candidate.size(); ++k) {
      const char c = candidate[k];
      const bool ok = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') ||
                      (c >= '0' && c <= '9') || c == '_';
      if (!ok) return false;
    }
    name = candidate;
    value.segs.clear();
    value.quoted = w.quoted;
    const std::string rest = first.substr(eq + 1);
    if (!rest.empty()) {
      Segment s;
      s.kind = Segment::Kind::Lit;
      s.text = rest;
      value.segs.push_back(std::move(s));
    }
    for (std::size_t k = 1; k < w.segs.size(); ++k) value.segs.push_back(w.segs[k]);
    return true;
  }

  static bool is_reserved(const Word& w) {
    auto lit = w.literal();
    return lit.has_value() && tables::is_reserved_word(*lit);
  }

  bool parse_script(unsigned depth, bool has_stop, Tok::Kind stop, Script& out) {
    if (depth > limits::kMaxNestDepth) return fail("nesting too deep");
    for (;;) {
      skip_newlines();
      if (eof() || at_stop(has_stop, stop)) break;

      Pipeline pl;
      if (!parse_pipeline(depth, pl)) return false;
      if (pl.cmds.empty()) return fail("empty command");
      out.pipelines.push_back(std::move(pl));

      if (at_separator()) { ++i; continue; }
      if (eof() || at_stop(has_stop, stop)) break;
      return fail("unexpected token after command");
    }
    return true;
  }

  bool parse_pipeline(unsigned depth, Pipeline& out) {
    for (;;) {
      Cmd c;
      bool produced = false;
      if (!parse_command(depth, c, produced)) return false;
      if (!produced) {
        if (!out.cmds.empty()) return fail("empty pipeline stage");
        return true;
      }
      out.cmds.push_back(std::move(c));
      if (!eof() && t[i].kind == Tok::Kind::Pipe) {
        ++i;
        skip_newlines();
        continue;
      }
      return true;
    }
  }

  bool parse_command(unsigned depth, Cmd& out, bool& produced) {
    produced = false;
    const Tok* p = peek();
    if (p != nullptr && (p->kind == Tok::Kind::LParen || p->kind == Tok::Kind::LBrace)) {
      const bool paren = p->kind == Tok::Kind::LParen;
      ++i;
      auto inner = std::make_shared<Script>();
      if (!parse_script(depth + 1, true, paren ? Tok::Kind::RParen : Tok::Kind::RBrace, *inner)) {
        return false;
      }
      const Tok* closer = peek();
      if (closer == nullptr ||
          closer->kind != (paren ? Tok::Kind::RParen : Tok::Kind::RBrace)) {
        return fail(paren ? "unclosed subshell" : "unclosed brace group");
      }
      ++i;
      out.kind = Cmd::Kind::Nested;
      out.nested = inner;
      produced = true;
      return true;
    }
    return parse_simple(depth, out, produced);
  }

  bool parse_simple(unsigned depth, Cmd& out, bool& produced) {
    SimpleCmd cmd;
    bool saw_any = false;

    for (;;) {
      const Tok* p = peek();
      if (p == nullptr) break;

      if (p->kind == Tok::Kind::Word) {
        // Function definition: NAME ( ) body
        const Tok* n1 = at(i + 1);
        const Tok* n2 = at(i + 2);
        if (cmd.words.empty() && n1 != nullptr && n1->kind == Tok::Kind::LParen &&
            n2 != nullptr && n2->kind == Tok::Kind::RParen) {
          auto lit = p->word.literal();
          if (!lit.has_value()) return fail("dynamic function name");
          const std::string name = *lit;
          i += 3;
          skip_newlines();
          Cmd body_cmd;
          bool body_produced = false;
          if (!parse_command(depth + 1, body_cmd, body_produced)) return false;
          if (!body_produced) return fail("function without a body");
          auto body = std::make_shared<Script>();
          if (body_cmd.kind == Cmd::Kind::Nested) {
            body = body_cmd.nested;
          } else {
            Pipeline pl;
            pl.cmds.push_back(std::move(body_cmd));
            body->pipelines.push_back(std::move(pl));
          }
          out.kind = Cmd::Kind::FuncDef;
          out.func_name = name;
          out.nested = body;
          produced = true;
          return true;
        }

        if (cmd.words.empty() && is_reserved(p->word)) {
          ++i;
          saw_any = true;
          continue;
        }

        if (cmd.words.empty()) {
          std::string name;
          Word value;
          if (try_split_assignment(p->word, name, value)) {
            cmd.assigns.emplace_back(name, value);
            ++i;
            saw_any = true;
            continue;
          }
        }

        cmd.words.push_back(p->word);
        ++i;
        saw_any = true;
        continue;
      }

      if (p->kind == Tok::Kind::Redir) {
        Redirect rd;
        rd.fd = p->fd;
        rd.op = p->op;
        ++i;
        const Tok* target = peek();
        if (target == nullptr || target->kind != Tok::Kind::Word) {
          return fail("redirection without a target");
        }
        rd.target = target->word;
        ++i;
        cmd.redirects.push_back(std::move(rd));
        saw_any = true;
        continue;
      }

      break;
    }

    if (!saw_any) return true;
    out.kind = Cmd::Kind::Simple;
    out.simple = std::move(cmd);
    produced = true;
    return true;
  }
};

}  // namespace

ParseResult parse(const std::vector<Tok>& toks) {
  ParseResult r;
  Parser p(toks);
  if (!p.parse_script(0, false, Tok::Kind::Word, r.script)) {
    r.ok = false;
    r.error = p.error;
    return r;
  }
  if (p.i < toks.size()) {
    r.ok = false;
    r.error = "unexpected trailing token";
    return r;
  }
  r.ok = true;
  return r;
}

ParseResult parse_source(const std::string& src) {
  LexResult lex = tokenize(src);
  if (!lex.ok) {
    ParseResult r;
    r.ok = false;
    r.error = lex.error;
    return r;
  }
  return parse(lex.toks);
}

}  // namespace agentgate

File: rust/crates/agentgate-core/src/parser.rs
//! Builds an AST from the token stream.
//!
//! The grammar is deliberately smaller than real POSIX shell. It captures
//! everything that determines *what will execute* -- command boundaries,
//! pipelines, subshells, brace groups, function definitions, assignments and
//! redirections -- and ignores everything that does not, such as the difference
//! between `;` and `&&`. For a safety analysis both branches of `&&` may run, so
//! collapsing them is the conservative choice.

use crate::lexer::{RedirOp, Tok, Word};
use crate::limits::MAX_NEST_DEPTH;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Redirect {
    pub fd: Option<i32>,
    pub op: RedirOp,
    pub target: Word,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct SimpleCmd {
    /// Leading `NAME=value` assignments.
    pub assigns: Vec<(String, Word)>,
    /// The command name and its arguments.
    pub words: Vec<Word>,
    pub redirects: Vec<Redirect>,
}

impl SimpleCmd {
    pub fn is_empty(&self) -> bool {
        self.words.is_empty() && self.redirects.is_empty() && self.assigns.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Cmd {
    Simple(SimpleCmd),
    /// A subshell `( ... )` or brace group `{ ...; }`.
    Nested(Script),
    /// `name() { ... }`
    FuncDef { name: String, body: Script },
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Pipeline {
    pub cmds: Vec<Cmd>,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Script {
    pub pipelines: Vec<Pipeline>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseError {
    pub message: String,
}

impl ParseError {
    fn new(m: &str) -> Self {
        ParseError { message: m.to_string() }
    }
}

struct P<'a> {
    t: &'a [Tok],
    i: usize,
}

impl<'a> P<'a> {
    fn peek(&self) -> Option<&'a Tok> {
        self.t.get(self.i)
    }
    fn bump(&mut self) -> Option<&'a Tok> {
        let t = self.t.get(self.i);
        if t.is_some() {
            self.i += 1;
        }
        t
    }
    fn at_separator(&self) -> bool {
        matches!(self.peek(), Some(Tok::Semi) | Some(Tok::Amp) | Some(Tok::Newline) | Some(Tok::AndIf) | Some(Tok::OrIf))
    }
    fn skip_newlines(&mut self) {
        while matches!(self.peek(), Some(Tok::Newline)) {
            self.i += 1;
        }
    }
}

/// Parse a token stream into a [`Script`].
pub fn parse(toks: &[Tok]) -> Result<Script, ParseError> {
    let mut p = P { t: toks, i: 0 };
    let s = parse_script(&mut p, 0, &[])?;
    if p.i < p.t.len() {
        return Err(ParseError::new("unexpected trailing token"));
    }
    Ok(s)
}

/// Parse commands until end of input or one of `stop`.
fn parse_script(p: &mut P, depth: u32, stop: &[Tok]) -> Result<Script, ParseError> {
    if depth > MAX_NEST_DEPTH {
        return Err(ParseError::new("nesting too deep"));
    }
    let mut script = Script::default();
    loop {
        p.skip_newlines();
        match p.peek() {
            None => break,
            Some(t) if stop.contains(t) => break,
            _ => {}
        }
        let pipe = parse_pipeline(p, depth)?;
        if pipe.cmds.is_empty() {
            return Err(ParseError::new("empty command"));
        }
        script.pipelines.push(pipe);
        if p.at_separator() {
            p.bump();
            continue;
        }
        match p.peek() {
            None => break,
            Some(t) if stop.contains(t) => break,
            Some(_) => return Err(ParseError::new("unexpected token after command")),
        }
    }
    Ok(script)
}

fn parse_pipeline(p: &mut P, depth: u32) -> Result<Pipeline, ParseError> {
    let mut pl = Pipeline::default();
    loop {
        let c = parse_command(p, depth)?;
        match c {
            Some(c) => pl.cmds.push(c),
            None => {
                // An empty stage is an error only when a pipe demanded one.
                if !pl.cmds.is_empty() {
                    return Err(ParseError::new("empty pipeline stage"));
                }
                return Ok(pl);
            }
        }
        if matches!(p.peek(), Some(Tok::Pipe)) {
            p.bump();
            p.skip_newlines();
            continue;
        }
        return Ok(pl);
    }
}

fn parse_command(p: &mut P, depth: u32) -> Result<Option<Cmd>, ParseError> {
    match p.peek() {
        Some(Tok::LParen) => {
            p.bump();
            let inner = parse_script(p, depth + 1, &[Tok::RParen])?;
            if !matches!(p.bump(), Some(Tok::RParen)) {
                return Err(ParseError::new("unclosed subshell"));
            }
            Ok(Some(Cmd::Nested(inner)))
        }
        Some(Tok::LBrace) => {
            p.bump();
            let inner = parse_script(p, depth + 1, &[Tok::RBrace])?;
            if !matches!(p.bump(), Some(Tok::RBrace)) {
                return Err(ParseError::new("unclosed brace group"));
            }
            Ok(Some(Cmd::Nested(inner)))
        }
        _ => parse_simple(p, depth),
    }
}

/// True when a literal word is a shell reserved word, which we skip so that
/// `if rm -rf /; then ...` still surfaces the `rm` command.
fn is_reserved(w: &Word) -> bool {
    match w.literal() {
        Some(l) => crate::tables::is_reserved_word(&l),
        None => false,
    }
}

/// Split `NAME=value` into its parts when the word begins with an assignment.
fn try_split_assignment(w: &Word) -> Option<(String, Word)> {
    use crate::lexer::Segment;
    if w.segs.is_empty() {
        return None;
    }
    let first = match &w.segs[0] {
        Segment::Lit(t) => t,
        _ => return None,
    };
    let eq = first.find('=')?;
    let name = &first[..eq];
    if name.is_empty() {
        return None;
    }
    let mut chars = name.chars();
    let c0 = chars.next()?;
    if !(c0.is_ascii_alphabetic() || c0 == '_') {
        return None;
    }
    if !chars.all(|c| c.is_ascii_alphanumeric() || c == '_') {
        return None;
    }
    let mut segs: Vec<Segment> = Vec::new();
    let rest = &first[eq + 1..];
    if !rest.is_empty() {
        segs.push(Segment::Lit(rest.to_string()));
    }
    segs.extend(w.segs[1..].iter().cloned());
    Some((name.to_string(), Word { segs, quoted: w.quoted }))
}

fn parse_simple(p: &mut P, depth: u32) -> Result<Option<Cmd>, ParseError> {
    let mut cmd = SimpleCmd::default();
    let mut saw_any = false;

    loop {
        match p.peek() {
            Some(Tok::Word(w)) => {
                // Function definition: NAME ( ) body
                if cmd.words.is_empty()
                    && matches!(p.t.get(p.i + 1), Some(Tok::LParen))
                    && matches!(p.t.get(p.i + 2), Some(Tok::RParen))
                {
                    let name = match w.literal() {
                        Some(l) => l,
                        None => return Err(ParseError::new("dynamic function name")),
                    };
                    p.i += 3;
                    p.skip_newlines();
                    let body = match parse_command(p, depth + 1)? {
                        Some(Cmd::Nested(s)) => s,
                        Some(other) => Script { pipelines: vec![Pipeline { cmds: vec![other] }] },
                        None => return Err(ParseError::new("function without a body")),
                    };
                    return Ok(Some(Cmd::FuncDef { name, body }));
                }

                if cmd.words.is_empty() && is_reserved(w) {
                    p.bump();
                    saw_any = true;
                    continue;
                }

                if cmd.words.is_empty() {
                    if let Some((name, value)) = try_split_assignment(w) {
                        cmd.assigns.push((name, value));
                        p.bump();
                        saw_any = true;
                        continue;
                    }
                }

                cmd.words.push(w.clone());
                p.bump();
                saw_any = true;
            }
            Some(Tok::Redir { fd, op }) => {
                let (fd, op) = (*fd, *op);
                p.bump();
                let target = match p.peek() {
                    Some(Tok::Word(w)) => {
                        let w = w.clone();
                        p.bump();
                        w
                    }
                    _ => return Err(ParseError::new("redirection without a target")),
                };
                cmd.redirects.push(Redirect { fd, op, target });
                saw_any = true;
            }
            _ => break,
        }
    }

    if !saw_any {
        return Ok(None);
    }
    if cmd.is_empty() {
        // Only reserved words were consumed, e.g. a bare `fi`.
        return Ok(Some(Cmd::Simple(cmd)));
    }
    Ok(Some(Cmd::Simple(cmd)))
}

/// Convenience: tokenize and parse in one step.
pub fn parse_source(src: &str) -> Result<Script, ParseError> {
    let toks = crate::lexer::tokenize(src).map_err(|e| ParseError { message: e.message })?;
    parse(&toks)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn first_cmd_name(s: &Script) -> String {
        match &s.pipelines[0].cmds[0] {
            Cmd::Simple(c) => c.words[0].literal().unwrap(),
            _ => panic!("expected a simple command"),
        }
    }

    #[test]
    fn parses_a_simple_command() {
        let s = parse_source("ls -la").unwrap();
        assert_eq!(s.pipelines.len(), 1);
        assert_eq!(first_cmd_name(&s), "ls");
    }

    #[test]
    fn separators_split_pipelines() {
        let s = parse_source("ls; rm -rf /").unwrap();
        assert_eq!(s.pipelines.len(), 2);
        let s = parse_source("ls\nrm -rf /").unwrap();
        assert_eq!(s.pipelines.len(), 2);
        let s = parse_source("true && rm -rf /").unwrap();
        assert_eq!(s.pipelines.len(), 2);
    }

    #[test]
    fn pipeline_stages_are_grouped() {
        let s = parse_source("curl http://x | sh").unwrap();
        assert_eq!(s.pipelines.len(), 1);
        assert_eq!(s.pipelines[0].cmds.len(), 2);
    }

    #[test]
    fn assignments_are_separated_from_words() {
        let s = parse_source("R=rm; $R -rf /").unwrap();
        match &s.pipelines[0].cmds[0] {
            Cmd::Simple(c) => {
                assert_eq!(c.assigns.len(), 1);
                assert_eq!(c.assigns[0].0, "R");
                assert_eq!(c.assigns[0].1.literal().unwrap(), "rm");
                assert!(c.words.is_empty());
            }
            _ => panic!("expected simple"),
        }
    }

    #[test]
    fn subshell_and_group_nest() {
        let s = parse_source("( rm -rf / )").unwrap();
        assert!(matches!(s.pipelines[0].cmds[0], Cmd::Nested(_)));
        let s = parse_source("{ rm -rf / ; }").unwrap();
        assert!(matches!(s.pipelines[0].cmds[0], Cmd::Nested(_)));
    }

    #[test]
    fn function_definition_captures_body() {
        let s = parse_source(":(){ :|:& };:").unwrap();
        match &s.pipelines[0].cmds[0] {
            Cmd::FuncDef { name, body } => {
                assert_eq!(name, ":");
                assert_eq!(body.pipelines[0].cmds.len(), 2);
            }
            other => panic!("expected FuncDef, got {other:?}"),
        }
    }

    #[test]
    fn redirects_attach_to_their_command() {
        let s = parse_source("echo hi > /dev/sda").unwrap();
        match &s.pipelines[0].cmds[0] {
            Cmd::Simple(c) => {
                assert_eq!(c.redirects.len(), 1);
                assert_eq!(c.redirects[0].target.literal().unwrap(), "/dev/sda");
            }
            _ => panic!("expected simple"),
        }
    }

    #[test]
    fn reserved_words_do_not_hide_commands() {
        let s = parse_source("if rm -rf /; then ls; fi").unwrap();
        assert_eq!(first_cmd_name(&s), "rm");
    }

    #[test]
    fn empty_and_malformed_inputs() {
        assert_eq!(parse_source("").unwrap().pipelines.len(), 0);
        assert_eq!(parse_source("   ").unwrap().pipelines.len(), 0);
        assert_eq!(parse_source("# comment").unwrap().pipelines.len(), 0);
        assert!(parse_source(";;;").is_err());
        assert!(parse_source("ls |").is_err());
        assert!(parse_source("( ls").is_err());
        assert!(parse_source("cat >").is_err());
    }

    #[test]
    fn trailing_separator_is_fine() {
        assert_eq!(parse_source("ls;").unwrap().pipelines.len(), 1);
        assert_eq!(parse_source("ls &").unwrap().pipelines.len(), 1);
    }
}
